### DISCUSSION TITLE
zh/tutorials/41-xdp-tcpdump/

### DISCUSSION THREAD

**Discussion by giscus** [eunomia-bpf/bpf-developer-tutorial#149] [General]
# zh/tutorials/41-xdp-tcpdump/

Unlock the potential of eBPF

https://eunomia.dev/zh/tutorials/41-xdp-tcpdump/

**Comment by kylecui** [comment]
a normal tcp header should be 20 bytes and it can be up to 60 bytes with TCP optional data. why the tcp header size is set to 32 here?

**Reply by yunwei37** [reply]
Good question. The `32` in this example is a fixed snapshot size, not the TCP header length calculated from the packet. It captures the 20-byte base header plus the first 12 bytes of any options, so it can truncate a valid TCP header, which may be as large as 60 bytes. A complete implementation should derive the length from `tcp->doff * 4`, validate it against the 20–60 byte range and `data_end`, then copy that bounded length.

**Comment by vigneshs-dev** [comment]
This works really fine.
